edx_exams/apps/access/views.py

```diff
@@ -0,0 +1,63 @@
+"""
+Exam Access Ticket Views
+"""
+from datetime import datetime, timedelta
+
+from django.http import HttpResponse
+from rest_framework import status
+from rest_framework.views import APIView
+from token_utils.sign import create_jwt
+
+from edx_exams.apps.core.models import ExamAttempt
+
+
+class ExamAccessTicketsView(APIView):
+    """
+    todo: add docs after impl is complete
+    """
+
+    # todo: should this be a get request?
+    def get(self, request):
+        user = request.user
+        exam = request.exam
+        claims = {}
+        claims.update(exam.course_id)
+        claims.update(exam.content_id)
+        exp_secs = 60
+
+        exam_attempt = ExamAttempt.get_latest_user_exam_attempt(user, exam)
+        # todo: should this be called ticket or token? For consistency?
+        cookie_name = "exam_access_ticket"
+
+        # If exam attempt exists for user, then grant exam access
+        if exam_attempt is not None:
+            # Grant access based on a short check-in window or the time to the end of the exam attempt,
+            # whichever is shorter.
+            if exam_attempt.start_time is not None and exam_attempt.allowed_time_limit_mins is not None:
+                end_time = exam_attempt.start_time + timedelta(minutes=exam_attempt.allowed_time_limit_mins)
+                secs_to_end = (end_time - datetime.now()).total_seconds()
+                if secs_to_end < exp_secs:
+                    exp_secs = secs_to_end
+            access_ticket = create_jwt(user.lms_user_id, exp_secs, claims)
+            # todo: What should be included in data? Since cookie includes access ticket.
+            response = HttpResponse("Created exam access ticket for latest attempt", 
+                                    status=status.HTTP_200_OK, 
+                                    data=[])
+            response.set_cookie(cookie_name, access_ticket, max_age=exp_secs)
+
+        # If exam is passed the due date, then grant exam access
+        elif exam.due_date is not None:
+            if datetime.now() > exam.due_date:
+                access_ticket = create_jwt(user.lms_user_id, exp_secs, claims)
```

Comment:
these few lines are a repeat of the case where the exam is in progress. Can you refactor a bit so we only need to build the ticket/response one time?

Reply:
Yeah, good call!